src/Microsoft.OData.Core/Json/JsonReaderExtensions.cs

```diff
@@ -391,8 +392,453 @@ internal static JsonNodeType ReadNext(this IJsonReader jsonReader)
+            object value = await jsonReader.GetValueAsync()
+                .ConfigureAwait(false);
+            return (string)value;
+        }
+
+        /// <summary>
+        /// Asynchronously reads the next node from the <paramref name="jsonReader"/>, verifies that it is a Property node and returns the property name.
+        /// </summary>
+        /// <param name="jsonReader">The <see cref="JsonReader"/> to read from.</param>
+        /// <returns>A task that represents the asynchronous read operation.
+        /// The value of the TResult parameter contains the property name of the property node read.</returns>
+        internal static async Task<string> ReadPropertyNameAsync(this IJsonReaderAsync jsonReader)
+        {
+            Debug.Assert(jsonReader != null, "jsonReader != null");
+
+            jsonReader.ValidateNodeType(JsonNodeType.Property);
+            string propertyName = await jsonReader.GetPropertyNameAsync()
+                .ConfigureAwait(false);
+            await jsonReader.ReadNextAsync()
+                .ConfigureAwait(false);
+
+            return propertyName;
+        }
+
+        /// <summary>
+        /// Asynchronously reads the next node from the <paramref name="jsonReader"/> and verifies that it is a PrimitiveValue node.
+        /// </summary>
+        /// <param name="jsonReader">The <see cref="JsonReader"/> to read from.</param>
+        /// <returns>A task that represents the asynchronous read operation.
+        /// The value of the TResult parameter contains the primitive value read from the reader.</returns>
+        internal static async Task<object> ReadPrimitiveValueAsync(this IJsonReaderAsync jsonReader)
+        {
+            Debug.Assert(jsonReader != null, "jsonReader != null");
+
+            object value = await jsonReader.GetValueAsync()
+                .ConfigureAwait(false);
+            await ReadNextAsync(jsonReader, JsonNodeType.PrimitiveValue)
+                .ConfigureAwait(false);
+
+            return value;
+        }
+
+        /// <summary>
+        /// Asynchronously reads the next node from the <paramref name="jsonReader"/> and verifies that it is a PrimitiveValue node of type string.
+        /// </summary>
+        /// <param name="jsonReader">The <see cref="JsonReader"/> to read from.</param>
+        /// <returns>A task that represents the asynchronous read operation.
+        /// The value of the TResult parameter contains the string value read from the reader;
+        /// throws an exception if no string value could be read.</returns>
+        internal static async Task<string> ReadStringValueAsync(this IJsonReaderAsync jsonReader)
+        {
+            Debug.Assert(jsonReader != null, "jsonReader != null");
+
+            object value = await jsonReader.ReadPrimitiveValueAsync()
+                .ConfigureAwait(false);
+
+            string stringValue = value as string;
+            if (value == null || stringValue != null)

```

Comment:
@odero There can be. The scenario below
```json
{"Prop":null}
```

Reply:
Hmm interesting. So in this case what would `stringValue` be?

Reply:
`stringValue` would be `null`. When reading, if what follows a property and the separator starts with `n`, then we try to read it as `null`; otherwise we throw an exception. For example we'd through an exception if we came across this: `{"Prop":nil}`. A non-null string however is expected to start with quotes, e.g., `{"Prop":"foo"}`. If we removed the `value == null` condition, then for `{"Prop":null}`, an exception would be thrown 4 lines below